### PR TITLE
Correctly handle blacklisted items in the template analysis

### LINKED ISSUE #584: Bindgen panics in Stylo on master
See https://bugzilla.mozilla.org/show_bug.cgi?id=1336013

```
 0:17.10 thread '<unnamed>' panicked at 'C'mon: Continue', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\result.rs:837
...
 0:17.10    7:     0x7ff63f888a2f - core::result::unwrap_failed<bindgen::parse::ParseError>
 0:17.10                         at C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\macros.rs:29
 0:17.10    8:     0x7ff63f8d46d0 - bindgen::ir::ty::Type::from_clang_ty
 0:17.10                         at S:\blue\third_party\rust\bindgen\src\ir\ty.rs:1024
...
0:17.11 thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Any', C:\bot\slave\stable-dist-rustc-win-msvc-64\build\src\libcore\result.rs:837
```

There's a couple places where I think we should stop asserting that we can always generate an IR item, and instead fall back to generating opaque items when we inevitably fail on something.

### PR DESCRIPTION
**Pull request by fitzgen**
The template analysis operates on whitelisted items, and uses our tracing
infrastructure to move between them. Usually, that means we can only reach other
whitelisted items by tracing, because the set of whitelisted items is the
transitive closure of all the items explicitly whitelisted. The exception is
when some type is explicitly blacklisted. It could still be reachable via
tracing from a whitelisted item, but is not considered whitelisted due to the
blacklisting.

The easy fix is to run the template analysis on the whole IR graph rather than
just the whitelisted set. This is an approximately one line change in the
analysis, however is not desirable due to performance concerns. The whole point
of whitelisting is that there may be *many* types in a header, but only a *few*
the user cares about, or there might be types that aren't explicitly needed and
that are too complicated for bindgen to handle generally (often in
`<type_traits>`). In these situations, we don't want to waste cycles or even
confuse ourselves by considering such types!

Instead, we keep the whitelisted item set around and check by hand whether any
given item is in it during the template type parameter analysis.

Additionally, we make the decision that blacklisted template definitions use all
of their type parameters. This seems like a reasonable choice because the type
will likely be ported to Rust manually by the bindgen user, and they will be
looking at the C++ definition with all of its type parameters. They can always
insert `PhantomData`s manually, so it also gives the most flexibility.

Fixes #584 

r? @emilio

### CODE REVIEW

**Comment by fitzgen** [comment]
With the changes in this PR, I can generate bindings for stylo successfully with libclang 3.9.

**Comment by emilio** [comment]
> With the changes in this PR, I can generate bindings for stylo successfully with libclang 3.9.

Nice! What about libclang 4.0/5.0? Does the regression affect that badly our analysis?

File: tests/headers/issue-584-stylo-template-analysis-panic.hpp [hunk 1]
```diff
@@ -0,0 +1,13 @@
+// bindgen-flags: --blacklist-type '^RefPtr$' --whitelist-function 'Servo_.*' --raw-line 'pub type RefPtr<T> = T;' -- -std=c++14
```
Comment (emilio) [inline tests/headers/issue-584-stylo-template-analysis-panic.hpp]: There's no need to `^$`, it's added implicitly.

**Comment by fitzgen** [comment]
> What about libclang 4.0/5.0? Does the regression affect that badly our analysis?

I haven't tested with other libclang versions.

The original test case within bindgen that used default template params looked like it came from gecko though, so if there is any instantiation that relies on the default type, it will fail. Note that this is orthogonal from the template parameter usage analysis. In this scenario, we simply aren't given the data we need, and this happens way before we do any template analysis.

**Comment by fitzgen** [comment]
@bors-servo r=emilio

Thanks for the review!

**Pull request merged**